File: src/ddc/command.py
"""
Command executor for ddcutil subprocess management.

This module provides the CommandExecutor class that handles execution of
ddcutil commands with proper timeout handling, retry logic, and output parsing.
"""

from __future__ import annotations

import logging
import re
import shlex
import subprocess
import time
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable

from src.ddc.exceptions import (
    CommandExecutionError,
    DDCNotAvailableError,
    PermissionError,
    TimeoutError,
)

logger = logging.getLogger(__name__)


@dataclass
class CommandResult:
    """
    Result of a ddcutil command execution.

    Attributes:
        success: Whether the command executed successfully
        return_code: The exit code of the command
        stdout: Standard output from the command
        stderr: Standard error from the command
        value: Extracted numeric value (for getvcp commands)
        command: The command that was executed
    """

    success: bool
    return_code: int
    stdout: str
    stderr: str
    value: int | None = None
    command: str = ""

    @property
    def error_message(self) -> str:
        """Get a formatted error message."""
        if self.success:
            return ""
        msg = f"Command failed with exit code {self.return_code}"
        if self.stderr:
            msg += f": {self.stderr.strip()}"
        return msg


class CommandExecutor:
    """
    Executor for ddcutil commands with retry logic and error handling.

    This class provides methods to execute ddcutil commands with proper
    timeout handling, retry logic for transient failures, and output parsing.
    """

    # Default configuration
    DEFAULT_TIMEOUT = 5.0  # seconds
    DEFAULT_MAX_RETRIES = 3
    DEFAULT_RETRY_DELAY = 0.1  # seconds
    DEFAULT_RETRY_BACKOFF = 2.0  # multiplier for exponential backoff

    # Patterns for parsing ddcutil output
    GETVCP_PATTERN = re.compile(r"VCP\s+(0x[0-9A-Fa-f]{2})\s+.*\s+([0-9]+)")
    DETECT_PATTERN = re.compile(r"I2C bus:\s*/dev/i2c-(\d+)")
    EDID_PATTERN = re.compile(
        r"Model:\s*(.+?)\s*\nSerial number:\s*(.+?)\s*\n"
        r"Manufactured:\s*(.+?)\s*\n"
        r"EDID version:\s*(.+?)\s*\n"
        r"Manufacturing ID:\s*(.+?)\s*\n"
        r"Display type:\s*(.+?)\s*\n"
    )

    def __init__(
        self,
        timeout: float = DEFAULT_TIMEOUT,
        max_retries: int = DEFAULT_MAX_RETRIES,
        retry_delay: float = DEFAULT_RETRY_DELAY,
        retry_backoff: float = DEFAULT_RETRY_BACKOFF,
    ) -> None:
        """
        Initialize the CommandExecutor.

        Args:
            timeout: Default timeout for command execution in seconds
            max_retries: Maximum number of retry attempts for transient failures
            retry_delay: Initial delay between retries in seconds
            retry_backoff: Multiplier for exponential backoff
        """
        self.timeout = timeout
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        self.retry_backoff = retry_backoff
        self._ddcutil_path: str | None = None

    def check_ddcutil_available(self) -> bool:
        """
        Check if ddcutil is available on the system.

        Returns:
            True if ddcutil is available, False otherwise
        """
        try:
            result = subprocess.run(
                ["ddcutil", "--version"],
                capture_output=True,
                text=True,
                timeout=2.0,
            )
            if result.returncode == 0:
                # Parse version to store path
                self._ddcutil_path = "ddcutil"
                logger.info(f"ddcutil found: {result.stdout.strip()}")
                return True
        except (subprocess.TimeoutExpired, FileNotFoundError):
            pass

        logger.warning("ddcutil not found on system")
        return False

    def _get_ddcutil_path(self) -> str:
        """
        Get the path to the ddcutil executable.

        Returns:
            Path to ddcutil

        Raises:
            DDCNotAvailableError: If ddcutil is not available
        """
        if self._ddcutil_path is None and not self.check_ddcutil_available():
            raise DDCNotAvailableError()
        return self._ddcutil_path or "ddcutil"

    def _execute(
        self,
        args: list[str],
        timeout: float | None = None,
    ) -> CommandResult:
        """
        Execute a ddcutil command without retry logic.

        Args:
            args: Command arguments (excluding ddcutil executable)
            timeout: Timeout in seconds, or None to use default

        Returns:
            CommandResult with execution details

        Raises:
            TimeoutError: If the command times out
        """
        command = [self._get_ddcutil_path()] + args
        timeout = timeout if timeout is not None else self.timeout
        command_str = " ".join(shlex.quote(arg) for arg in command)

        logger.debug(f"Executing: {command_str}")

        try:
            result = subprocess.run(
                command,
                capture_output=True,
                text=True,
                timeout=timeout,
            )

            return CommandResult(
                success=result.returncode == 0,
                return_code=result.returncode,
                stdout=result.stdout,
                stderr=result.stderr,
                command=command_str,
            )

        except subprocess.TimeoutExpired as e:
            logger.error(f"Command timed out after {timeout}s: {command_str}")
            raise TimeoutError(
                message=f"ddcutil command timed out after {timeout} seconds",
                timeout_seconds=timeout,
            ) from e
        except FileNotFoundError as e:
            logger.error("ddcutil executable not found")
            raise DDCNotAvailableError() from e
        except Exception as e:
            logger.error(f"Unexpected error executing command: {e}")
            raise CommandExecutionError(
                message=f"Unexpected error executing ddcutil command: {e}",
                command=command_str,
            ) from e

    def _should_retry(self, result: CommandResult) -> bool:
        """
        Determine if a failed command should be retried.

        Args:
            result: The command result to check

        Returns:
            True if the command should be retried, False otherwise
        """
        # Don't retry if it was a success
        if result.success:
            return False

        # Check for permission errors (don't retry)
        if "Permission denied" in result.stderr or "Permission" in result.stderr:
            return False

        # Check for monitor not found errors (don't retry)
        if "monitor not found" in result.stderr.lower():
            return False

        # Check for I2C bus errors (transient, should retry)
        if "i2c" in result.stderr.lower() or "ddc" in result.stderr.lower():
            return True

        # Check for timeout-related errors (should retry)
        if "timeout" in result.stderr.lower():
            return True

        # Default to not retry for unknown errors
        return False

    def execute(
        self,
        args: list[str],
        timeout: float | None = None,
        max_retries: int | None = None,
    ) -> CommandResult:
        """
        Execute a ddcutil command with retry logic.

        Args:
            args: Command arguments (excluding ddcutil executable)
            timeout: Timeout in seconds, or None to use default
            max_retries: Maximum number of retry attempts, or None to use default

        Returns:
            CommandResult with execution details

        Raises:
            DDCNotAvailableError: If ddcutil is not available
            PermissionError: If permission errors occur
            CommandExecutionError: If the command fails after all retries
        """
        max_retries = max_retries if max_retries is not None else self.max_retries
        retry_delay = self.retry_delay

        last_result: CommandResult | None = None

        for attempt in range(max_retries + 1):
            try:
                result = self._execute(args, timeout)

                # Check for permission errors
                if not result.success and "Permission denied" in result.stderr:
                    raise PermissionError(
                        message="Permission denied accessing I2C device",
                        stderr=result.stderr,
                    )

                # If successful or should not retry, return result
                if result.success or not self._should_retry(result):
                    return result

                last_result = result

            except TimeoutError:
                # Timeout errors are retryable
                if attempt < max_retries:
                    logger.warning(
                        f"Command timed out, retrying ({attempt + 1}/{max_retries})"
                    )
                    time.sleep(retry_delay)
                    retry_delay *= self.retry_backoff
                    continue
                raise

            except (PermissionError, DDCNotAvailableError):
                # These errors should not be retried
                raise

            except CommandExecutionError:
                # Unexpected errors should not be retried
                raise

        # If we get here, all retries failed
        if last_result:
            raise CommandExecutionError(
                message=f"Command failed after {max_retries} retries",
                command=last_result.command,
                exit_code=last_result.return_code,
                stderr=last_result.stderr,
            )

        raise CommandExecutionError(
            message="Command failed after retries (no result available)"
        )

    def parse_getvcp_output(self, output: str) -> tuple[int, int] | None:
        """
        Parse the output of a ddcutil getvcp command.

        Args:
            output: The stdout from getvcp command

        Returns:
            Tuple of (vcp_code, value) if parsing succeeds, None otherwise
        """
        # Try to match the standard ddcutil output format
        # Example: "VCP 10 (Brightness): current value = 75, maximum value = 100"
        match = re.search(r"VCP\s+(0x[0-9A-Fa-f]{2}|\d+)\s+.*current value\s*=\s*(\d+)", output)
        if match:
            code_str, value_str = match.groups()
            code = int(code_str, 0)  # Handles both hex and decimal
            value = int(value_str)
            return (code, value)

        # Alternative format: "0x10: 75 (0x4B)"
        match = re.search(r"(0x[0-9A-Fa-f]{2}):\s*(\d+)", output)
        if match:
            code = int(match.group(1), 0)
            value = int(match.group(2))
            return (code, value)

        return None

    def get_vcp(
        self,
        bus: int | None = None,
        vcp_code: int = 0x10,
        timeout: float | None = None,
    ) -> CommandResult:
        """
        Execute a ddcutil getvcp command.

        Args:
            bus: I2C bus number, or None for default monitor
            vcp_code: VCP code to query
            timeout: Timeout in seconds, or None to use default

        Returns:
            CommandResult with the parsed value in the 'value' field
        """
        args = ["getvcp", f"0x{vcp_code:02X}"]
        if bus is not None:
            args = [f"--bus={bus}"] + args

        result = self.execute(args, timeout)

        # Parse the output to extract the value
        if result.success:
            parsed = self.parse_getvcp_output(result.stdout)
            if parsed:
                result.value = parsed[1]

        return result

    def set_vcp(
        self,
        bus: int | None = None,
        vcp_code: int = 0x10,
        value: int = 50,
        timeout: float | None = None,
    ) -> CommandResult:
        """
        Execute a ddcutil setvcp command.

        Args:
            bus: I2C bus number, or None for default monitor
            vcp_code: VCP code to set
            value: Value to set
            timeout: Timeout in seconds, or None to use default

        Returns:
            CommandResult
        """
        args = ["setvcp", f"0x{vcp_code:02X}", str(value)]
        if bus is not None:
            args = [f"--bus={bus}"] + args

        return self.execute(args, timeout)

    def detect_monitors(self, timeout: float | None = None) -> CommandResult:
        """
        Execute a ddcutil detect command.

        Args:
            timeout: Timeout in seconds, or None to use default

        Returns:
            CommandResult with monitor detection output
        """
        args = ["detect"]
        return self.execute(args, timeout)

    def vcp_info(
        self,
        bus: int | None = None,
        timeout: float | None = None,
    ) -> CommandResult:
        """
        Execute a ddcutil vcpinfo command.

        Args:
            bus: I2C bus number, or None for default monitor
            timeout: Timeout in seconds, or None to use default

        Returns:
            CommandResult with VCP capability information
        """
        args = ["vcpinfo"]
        if bus is not None:
            args = [f"--bus={bus}"] + args

        return self.execute(args, timeout)

    def query_capabilities(
        self,
        bus: int | None = None,
        timeout: float | None = None,
    ) -> CommandResult:
        """
        Query monitor capabilities using ddcutil capabilities.

        Args:
            bus: I2C bus number, or None for default monitor
            timeout: Timeout in seconds, or None to use default

        Returns:
            CommandResult with capabilities output
        """
        args = ["capabilities"]
        if bus is not None:
            args = [f"--bus={bus}"] + args

        return self.execute(args, timeout)

File: twinkle-rust/src/ddc/command.rs
//! Command executor for ddcutil subprocess management.

use crate::ddc::error::{CommandExecutionError, DDCError, DDCResult};
use regex::Regex;
use std::process::Command;
use std::time::Duration;
use tokio::time::timeout as tokio_timeout;

/// Default timeout for command execution in seconds.
pub const DEFAULT_TIMEOUT_SECS: f64 = 5.0;

/// Default maximum number of retry attempts.
pub const DEFAULT_MAX_RETRIES: u32 = 3;

/// Default initial delay between retries in seconds.
pub const DEFAULT_RETRY_DELAY_SECS: f64 = 0.1;

/// Default multiplier for exponential backoff.
pub const DEFAULT_RETRY_BACKOFF: f64 = 2.0;

/// Result of a ddcutil command execution.
#[derive(Debug, Clone)]
pub struct CommandResult {
    /// Whether the command executed successfully
    pub success: bool,
    /// The exit code of the command
    pub return_code: i32,
    /// Standard output from the command
    pub stdout: String,
    /// Standard error from the command
    pub stderr: String,
    /// Extracted numeric value (for getvcp commands)
    pub value: Option<u16>,
    /// The command that was executed
    pub command: String,
}

impl CommandResult {
    /// Get a formatted error message.
    pub fn error_message(&self) -> String {
        if self.success {
            return String::new();
        }
        let mut msg = format!("Command failed with exit code {}", self.return_code);
        if !self.stderr.is_empty() {
            msg.push_str(&format!(": {}", self.stderr.trim()));
        }
        msg
    }
}

/// Executor for ddcutil commands with retry logic and error handling.
pub struct CommandExecutor {
    /// Default timeout for command execution in seconds
    timeout_secs: f64,
    /// Maximum number of retry attempts for transient failures
    max_retries: u32,
    /// Initial delay between retries in seconds
    retry_delay_secs: f64,
    /// Multiplier for exponential backoff
    retry_backoff: f64,
    /// Path to ddcutil executable
    ddcutil_path: Option<String>,
}

impl CommandExecutor {
    /// Create a new CommandExecutor with default settings.
    pub fn new() -> Self {
        Self {
            timeout_secs: DEFAULT_TIMEOUT_SECS,
            max_retries: DEFAULT_MAX_RETRIES,
            retry_delay_secs: DEFAULT_RETRY_DELAY_SECS,
            retry_backoff: DEFAULT_RETRY_BACKOFF,
            ddcutil_path: None,
        }
    }

    /// Create a new CommandExecutor with custom settings.
    pub fn with_config(
        timeout_secs: f64,
        max_retries: u32,
        retry_delay_secs: f64,
        retry_backoff: f64,
    ) -> Self {
        Self {
            timeout_secs,
            max_retries,
            retry_delay_secs,
            retry_backoff,
            ddcutil_path: None,
        }
    }

    /// Check if ddcutil is available on the system.
    /// This is an async function that uses timeout to prevent blocking.
    pub async fn check_ddcutil_available(&mut self) -> bool {
        tracing::info!("check_ddcutil_available() - Starting ddcutil --version command");
        
        let timeout_duration = Duration::from_secs_f64(self.timeout_secs);
        
        let result = tokio_timeout(timeout_duration, async {
            tokio::task::spawn_blocking(|| {
                Command::new("ddcutil")
                    .arg("--version")
                    .output()
            })
            .await
            .map_err(|e| DDCError::Other(format!("Task join error: {}", e)))?
            .map_err(|e| DDCError::Io(e))
        })
        .await;

        tracing::info!("check_ddcutil_available() - Command completed");
        match result {
            Ok(Ok(output)) => {
                if output.status.success() {
                    self.ddcutil_path = Some("ddcutil".to_string());
                    tracing::info!("ddcutil found: {}", String::from_utf8_lossy(&output.stdout).trim());
                    true
                } else {
                    tracing::warn!("ddcutil --version failed with exit code {:?}", output.status.code());
                    false
                }
            }
            Ok(Err(e)) => {
                tracing::warn!("ddcutil command failed: {}", e);
                false
            }
            Err(_) => {
                tracing::warn!("ddcutil --version timed out after {}s", self.timeout_secs);
                false
            }
        }
    }

    /// Get the path to the ddcutil executable.
    ///
    /// Returns an error if ddcutil is not available.
    async fn get_ddcutil_path(&mut self) -> DDCResult<String> {
        if self.ddcutil_path.is_none() && !self.check_ddcutil_available().await {
            return Err(DDCError::DDCNotAvailable);
        }
        Ok(self.ddcutil_path.clone().unwrap_or_else(|| "ddcutil".to_string()))
    }

    /// Execute a ddcutil command with timeout and retry logic.
    pub async fn execute(&mut self, args: &[&str]) -> DDCResult<CommandResult> {
        let ddcutil_path = self.get_ddcutil_path().await?;
        let command_str = format!("{} {}", ddcutil_path, args.join(" "));

        tracing::info!("execute() - Starting command: {} (timeout: {}s)", command_str, self.timeout_secs);
        let mut last_error = None;
        let mut delay = self.retry_delay_secs;

        for attempt in 0..=self.max_retries {
            if attempt > 0 {
                tracing::debug!("Retry attempt {} for command: {}", attempt, command_str);
                tokio::time::sleep(Duration::from_secs_f64(delay)).await;
                delay *= self.retry_backoff;
            }

            tracing::info!("execute() - Attempt {} for command: {}", attempt + 1, command_str);
            match self._execute_single(&ddcutil_path, args).await {
                Ok(result) => {
                    tracing::info!("execute() - Command completed with success={}", result.success);
                    if result.success {
                        return Ok(result);
                    }
                    // Check if this is a recoverable error
                    if self._is_recoverable(&result) {
                        last_error = Some(DDCError::CommandExecution(CommandExecutionError {
                            command: command_str.clone(),
                            exit_code: result.return_code,
                            stderr: result.stderr.clone(),
                        }));
                        continue;
                    }
                    return Err(DDCError::CommandExecution(CommandExecutionError {
                        command: command_str,
                        exit_code: result.return_code,
                        stderr: result.stderr,
                    }));
                }
                Err(e) => {
                    tracing::warn!("execute() - Command error: {}", e);
                    if is_timeout_error(&e) {
                        last_error = Some(e);
                        continue;
                    }
                    return Err(e);
                }
            }
        }

        Err(last_error.unwrap_or_else(|| {
            DDCError::CommandExecution(CommandExecutionError {
                command: command_str,
                exit_code: -1,
                stderr: "Max retries exceeded".to_string(),
            })
        }))
    }

    /// Execute a single command attempt with timeout.
    async fn _execute_single(&self, ddcutil_path: &str, args: &[&str]) -> DDCResult<CommandResult> {
        let timeout_duration = Duration::from_secs_f64(self.timeout_secs);

        // Clone to owned types for moving to spawn_blocking
        let ddcutil_path = ddcutil_path.to_string();
        let args: Vec<String> = args.iter().map(|s| s.to_string()).collect();
        let command_str = format!("{} {}", ddcutil_path, args.join(" "));

        let output = tokio_timeout(timeout_duration, async {
            tokio::task::spawn_blocking(move || {
                Command::new(&ddcutil_path)
                    .args(&args)
                    .output()
            })
            .await
            .map_err(|e| DDCError::Other(format!("Task join error: {}", e)))?
            .map_err(|e| DDCError::Io(e))
        })
        .await
        .map_err(|_| DDCError::Timeout(self.timeout_secs))??;

        let success = output.status.success();
        let return_code = output.status.code().unwrap_or(-1);
        let stdout = String::from_utf8_lossy(&output.stdout).to_string();
        let stderr = String::from_utf8_lossy(&output.stderr).to_string();

        Ok(CommandResult {
            success,
            return_code,
            stdout,
            stderr,
            value: None,
            command: command_str,
        })
    }

    /// Check if a command result indicates a recoverable error.
    fn _is_recoverable(&self, result: &CommandResult) -> bool {
        // Check for timeout-related errors
        if result.stderr.contains("timeout") || result.stderr.contains("timed out") {
            return true;
        }
        // Check for transient I2C errors
        if result.stderr.contains("I2C bus") || result.stderr.contains("DDC/CI") {
            return true;
        }
        false
    }

    /// Get a VCP value from a monitor.
    pub async fn get_vcp(&mut self, bus: i32, vcp_code: u8) -> DDCResult<CommandResult> {
        let args = &["getvcp", &format!("--bus={}", bus), &format!("0x{:02X}", vcp_code)];
        let mut result = self.execute(args).await?;

        // Parse the value from output
        if let Some(value) = self._parse_getvcp_output(&result.stdout) {
            result.value = Some(value);
        }

        Ok(result)
    }

    /// Set a VCP value on a monitor.
    pub async fn set_vcp(&mut self, bus: i32, vcp_code: u8, value: u16) -> DDCResult<CommandResult> {
        let args = &[
            "setvcp",
            &format!("--bus={}", bus),
            &format!("0x{:02X}", vcp_code),
            &value.to_string(),
        ];
        self.execute(args).await
    }

    /// Detect monitors on the system.
    pub async fn detect_monitors(&mut self) -> DDCResult<CommandResult> {
        tracing::info!("detect_monitors() - Starting ddcutil detect --brief command");
        let args = &["detect", "--brief"];
        let result = self.execute(args).await;
        tracing::info!("detect_monitors() - Command completed");
        result
    }

    /// Get EDID data for a monitor.
    pub async fn get_edid(&mut self, bus: i32) -> DDCResult<CommandResult> {
        let args = &["getedid", "--bus", &format!("{}", bus)];
        self.execute(args).await
    }

    /// Get capabilities for a monitor.
    pub async fn get_capabilities(&mut self, bus: i32) -> DDCResult<CommandResult> {
        let args = &["capabilities", "--bus", &format!("{}", bus)];
        self.execute(args).await
    }

    /// Parse the value from getvcp output.
    fn _parse_getvcp_output(&self, stdout: &str) -> Option<u16> {
        // Output format: "VCP 10 C 50 100" or "VCP 10 (Brightness): current value = 50, maximum = 100"
        let re = Regex::new(r"current value\s*=\s*(\d+)").ok()?;
        re.captures(stdout)?.get(1)?.as_str().parse().ok()
    }
}

impl Default for CommandExecutor {
    fn default() -> Self {
        Self::new()
    }
}

/// Check if an error is a timeout error.
fn is_timeout_error(err: &DDCError) -> bool {
    matches!(err, DDCError::Timeout(_))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_command_result_error_message() {
        let result = CommandResult {
            success: false,
            return_code: 1,
            stdout: String::new(),
            stderr: "Permission denied".to_string(),
            value: None,
            command: "ddcutil getvcp".to_string(),
        };
        assert_eq!(
            result.error_message(),
            "Command failed with exit code 1: Permission denied"
        );
    }

    #[test]
    fn test_parse_getvcp_output() {
        let executor = CommandExecutor::new();
        let output = "VCP 10 (Brightness): current value = 50, maximum = 100";
        assert_eq!(executor._parse_getvcp_output(output), Some(50));
    }
}
